### PR TITLE
Add e2e tests for consecutive changing runs

### PR DESCRIPTION
**Pull request by yguedidi**
second prerequisite for #3614

### CODE REVIEW

File: e2e/e2eTestChangingRunnerWithCache.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use Symfony\Component\Console\Command\Command;
+
+$projectRoot = __DIR__ .'/..';
+$rectorBin = $projectRoot . '/bin/rector';
+$autoloadFile = $projectRoot . '/vendor/autoload.php';
+
+// so we can use helper classes here
+require_once __DIR__ . '/../vendor/autoload.php';
+
+$e2eCommand = 'php '. $rectorBin .' process --no-ansi -a '. $autoloadFile;
```
Comment (yguedidi) [inline e2e/e2eTestChangingRunnerWithCache.php:20]: this new runner removed the `--dry-run` option

File: e2e/e2eTestChangingRunnerWithCache.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+if (isset($argv[1]) && $argv[1] === '-o') {
+    $expectedDiff = $argv[2];
+} else {
+    $expectedDiff = 'expected-output.diff';
+}
```
Comment (yguedidi) [inline e2e/e2eTestChangingRunnerWithCache.php:26]: and adds support for a specific expected output file

File: e2e/consecutive-changes/rector.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+use Rector\CodingStyle\Rector\Stmt\NewlineAfterStatementRector;
+use Rector\Config\RectorConfig;
+use Rector\EarlyReturn\Rector\If_\ChangeAndIfToEarlyReturnRector;
+
+return static function (RectorConfig $rectorConfig): void {
+    $rectorConfig->paths([
+        __DIR__ . '/src',
+    ]);
+
```
Comment (samsonasik) [inline e2e/consecutive-changes/rector.php]: FileCacheStorage need to be used like in consecutive dry run before 

https://github.com/rectorphp/rector-src/blob/dc57cffad71a875b8ccd11be265260754c47d32e/e2e/applied-rule-removed-node-with-cache/rector.php#L11

Comment (yguedidi) [reply e2e/consecutive-changes/rector.php]: good catch!

Comment (yguedidi) [reply e2e/consecutive-changes/rector.php]: I suffixed the folder with `-with-cache` to be explicit

**Comment by yguedidi** [comment]
@samsonasik fixed

**Comment by yguedidi** [comment]
@samsonasik Seeing that this PR make #3614 fails, I get a bit confused..
Then I thought more about it and it's quite logical, as after the fist run the the file get cached in that intermediate version of it, and so no consecutive changes in the second run as the file will be skipped.

But I'm wondering if it's really a use case we want that users have to run Rector multiple times to get the final fixed version of a fila?
I don't think so, me I'd like to run Rector once, and be confident that it mades all the needed changes.

@TomasVotruba @samsonasik what do you think?

**Comment by yguedidi** [comment]
@samsonasik made it a draft PR meanwhile we discuss the above point

**Comment by samsonasik** [comment]
Consecutive run can happen, when changed node just reprinted, and require the print to build the new type, and then, next run of Rector required to finalize it, in rector-src, that happen by rectify auto commit CI until no longer change.

**Comment by yguedidi** [comment]
I understand it can happen, but I believe from a user perspective it's not something expected.
I think we could find a solution to that issue so users to have to deal with that.
If this seems OK for you I can try to comme with a solution

**Comment by samsonasik** [comment]
Rector auto refresh type and continue to next rule by default, but on some cases, eg: add new line and spacing, that need to be re-read after.

That known limitation https://github.com/rectorphp/rector#known-drawbacks

Trust me, that expected to have multiple run.

**Comment by yguedidi** [comment]
I do trust you, I'm just wondering why can't we make that "second run" programatically in some optimized way, so it's not up to end users, as I think they will expect Rector to do its full job by one run.
Do you know if this need for a second run (because of the need to re-read after) comes only from code style rules like NewlineAfterStatementRector?
Because the Known Drawbacks you linked to explain that Rector is not about styling because based on AST, so I would propose to remove all code style rules from Rector, if they are the only onces requiring that second run.

**Comment by yguedidi** [comment]
Maybe you know why some code styling rectors were introduced in the first place?
Personally I'd favor caching and speed execution and leaving code style to tools dedicated to that, than supporting few code styling rectors with the drawback of a needed second run and no possibility to have a trustable cache

**Comment by samsonasik** [comment]
That's complex, it is not only that, that just an example, other factor can happen, eg: on multiple re-print consecutively , rector expected to stop, otherwise, that will be infinite loop.

**Comment by yguedidi** [comment]
> eg: on multiple re-print consecutively , rector expected to stop, otherwise, that will be infinite loop.

is this covered by I test? I may try something :)

**Comment by samsonasik** [comment]
Mostly on multiple rules usage, you can see tests/Issues directory

**Comment by yguedidi** [comment]
@samsonasik just opened #3672, it would be awesome if you can have a look and give your feedback :)

**Comment by yguedidi** [comment]
@samsonasik meanwhile the discussion get finalized in #3672 , let's move forward on this one as it's aligned with the current maintainers vision :)
If things change in the future we always can adapt this test

**Comment by yguedidi** [comment]
@samsonasik marked as ready for review, as I would like to advance on #3614 so that I can benefit from a reliable cache in my day to day work :)

**Comment by staabm** [comment]
fwiw, we also experienced problems with a stale cache in our daily work.

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @yguedidi

**Comment by yguedidi** [comment]
@staabm I'm on it! soon expect it to be fixed :)

**Comment by TomasVotruba** [comment]
I wonder if we can improve this experience 🙂

This stmts rule can have custom interface check, right?

**Comment by samsonasik** [comment]
@TomasVotruba this test make sure that multiple runs can still apply, on use case extract one statement to multi statements, then add new line between them, which the the line information is only can be fetched after rector done, which expected.

**Comment by yguedidi** [comment]
> This stmts rule can have custom interface check, right?

@TomasVotruba It's not clear to me what you mean here, sorry.

Would #3672 proposed PR be a better experience in your opinion? :)

**Pull request merged**